### PR TITLE
Zip Pair/Triple

### PR DESCRIPTION
**Pull request by stepango**

### CODE REVIEW

**Comment by marcinsus** [comment]
As for me, a useful feature. 👍

**Comment by thomasnield** [comment]
Might want to add a Pair extension for toMap() as well

**Comment by marcinsus** [comment]
@stepango two things: 
1. This types of method should be also in other method like latestFrom etc.
2. It is very nice to use but for me it is little inconsistent with base rxjava which requires zipper to connect two observable. I am not sure if we change this behaviour some rx users can be little confused.
 
Another way to archive it is create helper method something like:
` fun <T, U> pairZipper(): (T, U) -> Pair<T, U> = { first: T, second: U ->
        first to second
    }
`

**Comment by stepango** [comment]
@marcinsus `pairZipper` does not make any sense, you could just use `::Pair` with one of rxKotlin's extension methods if you like
Agreed with you on 1. Regarding 2nd - I think of it like a default zipper. On my latest project, I found that in 90% Pair/Triple all I needed from zip/combineLatest methods.

**Comment by marcinsus** [comment]
@stepango are you sure `::Pair` is possible? I had problem with type inference when i try refactor something like that :
```
 Observable.just(3) .zipWith(Observable.just(""), { a, b -> a to b})
```
Also in my case the most of usage when zip 2 observable is create pair so for me default zipper is ok. I am not sure but maybe default parameter instead method will be ok?

**Comment by stepango** [comment]
@marcinsus `Observables.zip(Observable.just(1), Observable.just(1), ::Pair)`

**Comment by marcinsus** [comment]
@stepango yeah, Observables.zip (and similar object method) work without problem but extension method like zipWith method has name conflict with rxjava method and ::Pair does not work.

**Comment by stepango** [comment]
@marcinsus you are right about `::Pair` looks like a bug that type inference does not work for method references if method has overload with same amount of params https://youtrack.jetbrains.com/issue/KT-18080

**Comment by thomasnield** [comment]
Is there a `combineLatestWith()` natively with RxJava? I know there is a `withLatestFrom()` which is similar but only triggers from the invoked Observable.

**Comment by stepango** [comment]
@thomasnield no, there is no `combineLatestWith` and `withLatestFrom` behave differently. I don't think it's necessary to rename `combineLatest` to `combineLatestWith` it could confuse people

File: src/main/kotlin/io/reactivex/rxkotlin/Observables.kt [hunk 1]
```diff
@@ -141,8 +154,49 @@ inline fun <T, T1, T2, T3, T4, R> Observable<T>.withLatestFrom(o1: ObservableSou
         = withLatestFrom(o1, o2, o3, o4, Function5<T, T1, T2, T3, T4, R> { t, t1, t2, t3, t4 -> combiner.invoke(t, t1, t2, t3, t4) })
 
 /**
+ * Shortcut for [Observable.combineLatest] with two args
```
Comment (thomasnield) [inline src/main/kotlin/io/reactivex/rxkotlin/Observables.kt:157]: I propose removing `combineLatestWith()` since this is not natively available in RxJava.

**Review by thomasnield — CHANGES_REQUESTED** [review]
Remove combineLatestWith()

**Comment by thomasnield** [comment]
Okay, i see there is a `combineLatest()` already in this library that I must have overlooked. I don't agree with it being there, even if it's name is changed to `combineLatestWith()`. I do think a factory overload for `combineLatestWith` in `Observables` and `Flowables` that yield `Pair` and `Triple` would be great though.

**Comment by thomasnield** [comment]
And again, the reason is because it does not exist in the original RxJava implementation as an operator, only a factory.

**Comment by thomasnield** [comment]
I went ahead and cherrypicked the ideas from this PR  to implement following:

```
Observables#zip(): Pair<T1,T2>
Observables#zip(): Triple<T1,T2,T3>
Observables#combineLatest(): Pair<T1,T2>
Observables#combineLatest(): Triple<T1,T2,T3>
Observable.zipWith(): Pair<T1,T2>
Observable.withLatestFrom(): Pair<T1,T2>
```

I'm deprecating `combineLatest()` as an operator, and not going forward with `combineLatestWith()` although that would have been a more sensible name for it. 

Feel free to reopen this PR if I missed anything or you feel strongly about any of the above changes.

**Pull request closed without merging**